python_modules/dagster-test/dagster_test/toys/user_computed_data_versions/__init__.py

```diff
@@ -0,0 +1,129 @@
+indicating whether a memoized value was used. This information is passed to the Dagster framework by
+returning a `Nothing` `Output`.
+"""
+import json
+import warnings
+from typing import cast
+
+from dagster import (
+    AssetKey,
+    AssetsDefinition,
+    DataProvenance,
+    DataVersion,
+    Definitions,
+    ExperimentalWarning,
+    In,
+    Nothing,
+    OpDefinition,
+    OpExecutionContext,
+    Out,
+    Output,
+    observable_source_asset,
+)
+from dagster._core.definitions.unresolved_asset_job_definition import define_asset_job
+
+from .external_system import ExternalSystem
+
+warnings.filterwarnings("ignore", category=ExperimentalWarning)
+
+EXTERNAL_SYSTEM = ExternalSystem()
+
+
+def external_asset(key: AssetKey):
+    """Factory to build an `AssetsDefinition` object to represent an asset in an external system.
+
+    The factory queries the external system for dependency and code version info. The op attached
+    to the `AssetsDefinition` forwards materialization requests to the external system, sending an
+    asset key and provenance info for the last recorded materialization of the asset. The external
+    system responds with a data version and a flag indicating whether a memoized value was used for
+    the materialization (i.e. whether recomputation actually occurred). This is passed to the
+    dagster framework by packaging it on a returned `Output` object. The `Output` returns `Nothing`
+    because Dagster is not handling passing data in between ops in this scenario.
+
+    Args:
+        key (AssetKey):
+            The asset key for the represented asset.
+
+    Returns (AssetsDefinition):
+        An `AssetsDefinition` object representing the asset in the external system.
+    """
+
+    def fn(context: OpExecutionContext, *args):
+        # Keep the temp storage directory used by the backend to the current instance.
+        EXTERNAL_SYSTEM.set_storage_directory(context.instance.storage_directory())
+
+        provenance = context.get_asset_provenance(key)
+        provenance_json = serialize_provenance(provenance) if provenance else None
+        data_version, is_memoized = EXTERNAL_SYSTEM.materialize(
+            key.to_user_string(),
+            provenance_json,
+        )
+        if is_memoized:
+            context.log.info(f"Used memoized value for {key.to_user_string()}")
+        yield Output(
+            None, data_version=DataVersion(data_version), metadata={"is_memoized": is_memoized}
+        )
+
+    deps = [AssetKey(k) for k in EXTERNAL_SYSTEM.get_dependencies(key.to_user_string())]
+    keys_by_input_name = {k.path[-1]: k for k in deps}
+    keys_by_output_name = {"result": key}
+    op_def = OpDefinition(
+        name=key.path[-1],
+        ins={k.to_user_string(): In(cast(type, Nothing)) for k in deps},
+        outs={"result": Out(Nothing)},
+        code_version=EXTERNAL_SYSTEM.get_code_version(key.to_user_string()),
```

Comment:
also do not think the fetching code versions should be the responsibility of "external system"